src/EventLoop.cpp

```diff
@@ -171,18 +190,73 @@ auto EventLoop::StartRequest(
     if (m_is_stopping) {
         return false;
     }
-
     // We'll prepare now since it won't block the event loop thread.
     request->prepareForPerform();
     {
         std::lock_guard<std::mutex> guard { m_pending_requests_lock };
         m_pending_requests.emplace_back(std::move(request));
     }
     uv_async_send(&m_async);
-
     return true;
 }
 
+auto EventLoop::stopTimedOutRequests() -> void
+{
+    // If the wrappers multiset is empty, then we don't have anything to stop
+    if (m_response_wait_time_wrappers.empty())
+    {
+        return;
+    }
+
+    // Get the current timepoint from libuv (it caches the current timepoint at the beginning
+    // of every cycle through the event loop, so this should not be expensive).
+    auto now = uv_now(m_loop);
+
+    for (
+        auto current_wrapper = m_response_wait_time_wrappers.begin();
+        current_wrapper != m_response_wait_time_wrappers.end() && current_wrapper->GetData().m_timeout_time <= now;
+        // We increment current_wrapper by setting it to the iterator returned from erasing the current one
+    )
+    {
+        // Get the shared pointer out of the ResponseWaitTimeWrapper and into a unique pointer so it cleans
+        // itself up at the end of this method.
+        auto& data = current_wrapper->GetData();
+        auto shared_request = std::unique_ptr<std::shared_ptr<SharedRequest>>(data.m_shared_request_ptr_pointer);
+        
+        // Call onComplete on the underlying Request object, copying in the shared_pointer
+        // so we get a correct reference count.
+        (**shared_request).GetAsReference().onComplete(*this, *shared_request, true);
+        
+        // Update current_wrapper with the result of removing the wrapper from the multiset
+        current_wrapper = removeTimeoutByIterator(current_wrapper);
+    }
+    
+    // If there are still items in the multiset, get the first item and use its time to reset the request timer.
+    if (!m_response_wait_time_wrappers.empty())
+    {
+        auto& request_time = m_response_wait_time_wrappers.begin()->GetData().m_timeout_time;
+        uv_timer_stop(&m_response_timer);
+    
+        uv_timer_start(
+            &m_response_timer,
+            on_response_wait_time_expired_callback,
+            request_time - now,
+            0);
+    }
+}
+
+auto EventLoop::removeTimeoutByIterator(std::multiset<ResponseWaitTimeWrapper>::iterator request_to_remove) -> std::multiset<ResponseWaitTimeWrapper>::iterator
+{
+    // Get the shared pointer to SharedRequest into a unique pointer so it cleans itself up.
```

Comment:
initially this looks like a double delete due to the unique_ptr acquired in stopTimedOutRequests
I think there should be a comment clearly indicating that you're creating a unique pointer to a NEW shared_ptr copy and what your intention is

Reply:
Good point -- getting a reference to the underlying structure instead of creating a copy of the shared pointer.